charms/istio-gateway/tests/test_charm.py

```diff
@@ -21,95 +15,70 @@ def test_no_kind(harness):
+    helpers.begin_noop(harness)
 
     assert harness.charm.model.unit.status == BlockedStatus('Waiting for istio-pilot relation')
 
 
-def test_kind_ingress(harness, mocker):
-    run = mocker.patch('subprocess.run')
-    harness.set_leader(True)
-
-    harness.update_config({'kind': 'ingress'})
-    harness.add_oci_resource(
-        "noop",
-        {
-            "registrypath": "",
-            "username": "",
-            "password": "",
-        },
-    )
-    rel_id = harness.add_relation("istio-pilot", "app")
-
-    harness.add_relation_unit(rel_id, "app/0")
-    data = {"service-name": "service-name", "service-port": '6666'}
-    harness.update_relation_data(
-        rel_id,
-        "app",
-        {"_supported_versions": "- v1", "data": yaml.dump(data)},
-    )
-
-    harness.begin_with_initial_hooks()
-
-    container = harness.model.unit.get_container('noop')
-    harness.charm.on['noop'].pebble_ready.emit(container)
-
-    assert len(run.call_args_list) == 4
-
-    expected_args = (['./kubectl', 'apply', '-f-'],)
-    expected_input = list(yaml.safe_load_all(open('tests/ingress-example.yaml')))
-
-    for call in run.call_args_list:
-        actual_input = list(yaml.safe_load_all(call.kwargs['input']))
-        assert call.args == expected_args
-        assert expected_input == actual_input
-
-    assert harness.charm.model.unit.status == ActiveStatus('')
-
-
-def test_kind_egress(harness, mocker):
-    run = mocker.patch('subprocess.run')
-    harness.set_leader(True)
-
-    harness.update_config({'kind': 'egress'})
-    harness.add_oci_resource(
-        "noop",
-        {
-            "registrypath": "",
-            "username": "",
-            "password": "",
-        },
-    )
-    rel_id = harness.add_relation("istio-pilot", "app")
-
-    harness.add_relation_unit(rel_id, "app/0")
-    data = {"service-name": "service-name", "service-port": '6666'}
-    harness.update_relation_data(
-        rel_id,
-        "app",
-        {"_supported_versions": "- v1", "data": yaml.dump(data)},
-    )
-
-    harness.begin_with_initial_hooks()
-
-    container = harness.model.unit.get_container('noop')
-    harness.charm.on['noop'].pebble_ready.emit(container)
-
-    assert len(run.call_args_list) == 4
-
-    expected_args = (['./kubectl', 'apply', '-f-'],)
-    expected_input = list(yaml.safe_load_all(open('tests/egress-example.yaml')))
-
-    for call in run.call_args_list:
-        actual_input = list(yaml.safe_load_all(call.kwargs['input']))
-        assert call.args == expected_args
-        assert expected_input == actual_input
-
-    assert harness.charm.model.unit.status == ActiveStatus('')
+def test_install_apply(configured_harness, kind, mocked_client, helpers):
+    actual_objects = []
+    expected_objects = list(yaml.safe_load_all(open(f'tests/{kind}-example.yaml')))
+
+    # The install method is invoked multiple times, and the apply method is called for every object
```

Comment:
i'm not opposed to testing with mocks this way.  
more often i see the patterns [here](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called)

```python
mocked_client.return_value.apply.assert_called_with(...)
mocked_client.return_value.apply.assert_not_called(...)
mocked_client.return_value.apply.assert_has_calls(...)
```

etc.  Any reason you went about iterating against the call list?

Reply:
yeah `assert_has_calls` is the closest -- but t just ensures that a subset of the calls were made in a specified order